source/nvda.pyw

```diff
@@ -213,12 +230,83 @@ def getDesktopName():
 	ctypes.windll.user32.GetUserObjectInformationW(desktop, UOI_NAME, ctypes.byref(name), ctypes.sizeof(name), None)
 	return name.value
 
-#Ensure multiple instances are not fully started by using a mutex
-ERROR_ALREADY_EXISTS=0XB7
-desktopName=getDesktopName()
-mutex=ctypes.windll.kernel32.CreateMutexW(None,True,u"Local\\NVDA_%s"%desktopName)
-if not mutex or ctypes.windll.kernel32.GetLastError()==ERROR_ALREADY_EXISTS:
-	if mutex: ctypes.windll.kernel32.CloseHandle(mutex)
+
+# Ensure multiple instances are not fully started by using a mutex
+desktopName = getDesktopName()
+log.info(f"DesktopName: {desktopName}")
+
+
+def _acquireMutex(_desktopName: str) -> typing.Optional[wintypes.HANDLE]:
+	# From MS docs; "Multiple processes can have handles of the same mutex object"
+	# > Two or more processes can call CreateMutex to create the same named mutex.
+	# > The first process actually creates the mutex, and subsequent processes with sufficient access rights
+	# > simply open a handle to the existing mutex.
+	# > This enables multiple processes to get handles of the same mutex, while relieving the user of the
+	# > responsibility of ensuring that the creating process is started first.
+	# > When using this technique, you should set the bInitialOwner flag to FALSE; otherwise, it can be difficult
+	# > to be certain which process has initial ownership.
+	# > https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createmutexw
+	_mutex = ctypes.windll.kernel32.CreateMutexW(
+		None,  # lpMutexAttributes,
+		# Don't take initial ownership, use wait to acquire ownership instead.
+		# Allows waiting for a prior process to finish exiting.
+		False,  # bInitialOwner
+		f"Local\\NVDA_{_desktopName}"  # lpName
+	)
+	createMutexResult = ctypes.windll.kernel32.GetLastError()
+	if not _mutex:
+		log.error(f"Unable to create mutex, last error: {createMutexResult}")
+		raise winUser.WinError(createMutexResult)
+	else:
+		if createMutexResult == winKernel.ERROR_ALREADY_EXISTS:
+			log.debug("Waiting for prior NVDA to finish exiting")
+		# We didn't ask to be the initial owner,
+		waitResult = winKernel.waitForSingleObject(
+			_mutex,  # hHandle
+			2000  # dwMilliseconds
+		)
+
+		log.debug(f"Wait result: {waitResult}")
+		if winKernel.WAIT_OBJECT_0 == waitResult:
+			log.info("Prior NVDA has finished exiting")
+			return _mutex  # mutex ownership acquired
+		elif winKernel.WAIT_ABANDONED == waitResult:
+			log.error(
+				"Prior NVDA exited without releasing mutex, taking ownership."
+				" Note: Restarting your system is recommended."
+				" This error indicates that NVDA previously did not exit correctly or was terminated"
+				" (perhaps by the task manager)."
+				" Your system is in an indeterminate state."

```

Comment:
This message really botheres me. I think the reboot part is enough. The system being in an indeterminant state may scare some beginner users more than necesary.

Reply:
It's true though right, NVDA crashing could leave the system in an unknown state. The word choice was intended to scare users into restarting. Perhaps it would be more helpful (at least for developers) if the key unknowns were listed?

Reply:
I've removed this line from the message.